File: PyLibs/src/Libs/FrameHandle/TimerItemBase.py
# -*- coding: utf-8 -*-

from Libs.Core.GObject import GObject;
from Libs.FrameHandle.TimerFunctionObject import TimerFunctionObject;

'''
@brief 定时器，这个是不断增长的
'''

class TimerItemBase(GObject):

    def __init__(self):
        super(TimerItemBase, self).__init__();
        
        self.mTypeId = "TimerItemBase";
        
        self.mInternal = 1;             # 定时器间隔
        self.mTotalTime = 1;            # 总共定时器时间
        self.mCurRunTime = 0;           # 当前定时器运行的时间
        self.mCurCallTime = 0;          # 当前定时器已经调用的时间
        self.mIsInfineLoop = False;      # 是否是无限循环
        self.mIntervalLeftTime = 0;     # 定时器调用间隔剩余时间
        self.mTimerDisp = TimerFunctionObject();    # 定时器分发
        self.mIsDisposed = False;         # 是否已经被释放
        self.mIsContinuous = False;      # 是否是连续的定时器


    def setFuncObject(self, handle):
        self.mTimerDisp.setFuncObject(handle);


    def setTotalTime(self, value):
        self.mTotalTime = value;


    def getRunTime(self):
        return self.mCurRunTime;


    def getCallTime(self):
        return self.mCurCallTime;


    def getLeftRunTime(self):
        return self.mTotalTime - self.mCurRunTime;


    def getLeftCallTime(self):
        return self.mTotalTime - self.mCurCallTime;


    # 在调用回调函数之前处理
    def onPreCallBack(self):
        pass;


    def OnTimer(self, delta):
        if (self.mIsDisposed):
            return;

        self.mCurRunTime += delta;
        if (self.mCurRunTime > self.mTotalTime):
            self.mCurRunTime = self.mTotalTime;

        self.mIntervalLeftTime += delta;

        if (self.mIsInfineLoop):
            self.checkAndDisp();
        else:
            if (self.mCurRunTime >= self.mTotalTime):
                self.disposeAndDisp();
            else:
                self.checkAndDisp();


    def disposeAndDisp(self):
        if (self.mIsContinuous):
            self.continueDisposeAndDisp();
        else:
            self.discontinueDisposeAndDisp();


    def continueDisposeAndDisp(self):
        self.mIsDisposed = True;

        while (self.mIntervalLeftTime >= self.mInternal and self.mCurCallTime < self.mTotalTime):
            self.mCurCallTime = self.mCurCallTime + self.mInternal;
            self.mIntervalLeftTime = self.mIntervalLeftTime - self.mInternal;
            self.onPreCallBack();

            if (self.mTimerDisp.isValid()):
                self.mTimerDisp.call(self);


    def discontinueDisposeAndDisp(self):
        self.mIsDisposed = True;
        self.mCurCallTime = self.mTotalTime;
        self.onPreCallBack();

        if (self.mTimerDisp.isValid()):
            self.mTimerDisp.call(self);


    def checkAndDisp(self):
        if(self.mIsContinuous):
            self.continueCheckAndDisp();
        else:
            self.discontinueCheckAndDisp();

    # 连续的定时器
    def continueCheckAndDisp(self):
        while (self.mIntervalLeftTime >= self.mInternal):
            # 这个地方 mCurCallTime 肯定会小于 mTotalTime，因为在调用这个函数的外部已经进行了判断
            self.mCurCallTime = self.mCurCallTime + self.mInternal;
            self.mIntervalLeftTime = self.mIntervalLeftTime - self.mInternal;
            self.onPreCallBack();

            if (self.mTimerDisp.isValid()):
                self.mTimerDisp.call(self);



    # 不连续的定时器
    def discontinueCheckAndDisp(self):
        if (self.mIntervalLeftTime >= self.mInternal):
            # 这个地方 mCurCallTime 肯定会小于 mTotalTime，因为在调用这个函数的外部已经进行了判断
            self.mCurCallTime = self.mCurCallTime + (((int)(self.mIntervalLeftTime / self.mInternal)) * self.mInternal);
            self.mIntervalLeftTime = self.mIntervalLeftTime % self.mInternal;   # 只保留余数
            self.onPreCallBack();

            if (self.mTimerDisp.isValid()):
                self.mTimerDisp.call(self);


    def reset(self):
        self.mCurRunTime = 0;
        self.mCurCallTime = 0;
        self.mIntervalLeftTime = 0;
        self.mIsDisposed = False;


    def setClientDispose(self):
        pass;


    def IsClientDispose(self):
        return False;

File: JSLib/src/FrameHandle/TimerItemBase.js
using LuaInterface;
using System;

namespace SDK.Lib
{
    /**
     * @brief 定时器，这个是不断增长的
     */
    public class TimerItemBase : IDelayHandleItem, IDispatchObject
    {
        public float m_internal;        // 定时器间隔
        public float m_totalTime;       // 总共定时器时间
        public float m_curRunTime;      // 当前定时器运行的时间
        public float m_curCallTime;     // 当前定时器已经调用的时间
        public bool m_bInfineLoop;      // 是否是无限循环
        public float m_intervalLeftTime;     // 定时器调用间隔剩余时间
        public TimerFunctionObject m_timerDisp;       // 定时器分发
        public bool m_disposed;             // 是否已经被释放
        public bool m_bContinuous;          // 是否是连续的定时器

        public TimerItemBase()
        {
            m_internal = 1;
            m_totalTime = 1;
            m_curRunTime = 0;
            m_curCallTime = 0;
            m_bInfineLoop = false;
            m_intervalLeftTime = 0;
            m_timerDisp = new TimerFunctionObject();
            m_disposed = false;
            m_bContinuous = false;
        }

        public void setFuncObject(Action<TimerItemBase> handle)
        {
            m_timerDisp.setFuncObject(handle);
        }

        virtual public void setTotalTime(float value)
        {
            this.m_totalTime = value;
        }

        virtual public float getRunTime()
        {
            return this.m_curRunTime;
        }

        virtual public float getCallTime()
        {
            return this.m_curCallTime;
        }

        virtual public float getLeftRunTime()
        {
            return this.m_totalTime - this.m_curRunTime;
        }

        virtual public float getLeftCallTime()
        {
            return this.m_totalTime - this.m_curCallTime;
        }

        // 在调用回调函数之前处理
        protected virtual void onPreCallBack()
        {

        }

        public virtual void OnTimer(float delta)
        {
            if (m_disposed)
            {
                return;
            }

            m_curRunTime += delta;
            if (m_curRunTime > m_totalTime)
            {
                m_curRunTime = m_totalTime;
            }
            m_intervalLeftTime += delta;

            if (m_bInfineLoop)
            {
                checkAndDisp();
            }
            else
            {
                if (m_curRunTime >= m_totalTime)
                {
                    disposeAndDisp();
                }
                else
                {
                    checkAndDisp();
                }
            }
        }

        public virtual void disposeAndDisp()
        {
            if (this.m_bContinuous)
            {
                this.continueDisposeAndDisp();
            }
            else
            {
                this.discontinueDisposeAndDisp();
            }
        }

        protected void continueDisposeAndDisp()
        {
            this.m_disposed = true;

            while (this.m_intervalLeftTime >= this.m_internal && this.m_curCallTime < this.m_totalTime)
            {
                this.m_curCallTime = this.m_curCallTime + this.m_internal;
                this.m_intervalLeftTime = this.m_intervalLeftTime - this.m_internal;
                this.onPreCallBack();

                if (this.m_timerDisp.isValid())
                {
                    this.m_timerDisp.call(this);
                }
            }
        }

        protected void discontinueDisposeAndDisp()
        {
            m_disposed = true;
            m_curCallTime = m_totalTime;
            this.onPreCallBack();

            if (m_timerDisp.isValid())
            {
                m_timerDisp.call(this);
            }
        }

        public virtual void checkAndDisp()
        {
            if(m_bContinuous)
            {
                continueCheckAndDisp();
            }
            else
            {
                discontinueCheckAndDisp();
            }
        }

        // 连续的定时器
        protected void continueCheckAndDisp()
        {
            while (m_intervalLeftTime >= m_internal)
            {
                // 这个地方 m_curCallTime 肯定会小于 m_totalTime，因为在调用这个函数的外部已经进行了判断
                m_curCallTime = m_curCallTime + m_internal;
                m_intervalLeftTime = m_intervalLeftTime - m_internal;
                this.onPreCallBack();

                if (m_timerDisp.isValid())
                {
                    m_timerDisp.call(this);
                }
            }
        }

        // 不连续的定时器
        protected void discontinueCheckAndDisp()
        {
            if (m_intervalLeftTime >= m_internal)
            {
                // 这个地方 m_curCallTime 肯定会小于 m_totalTime，因为在调用这个函数的外部已经进行了判断
                m_curCallTime = m_curCallTime + (((int)(m_intervalLeftTime / m_internal)) * m_internal);
                m_intervalLeftTime = m_intervalLeftTime % m_internal;   // 只保留余数
                this.onPreCallBack();

                if (m_timerDisp.isValid())
                {
                    m_timerDisp.call(this);
                }
            }
        }

        public virtual void reset()
        {
            m_curRunTime = 0;
            m_curCallTime = 0;
            m_intervalLeftTime = 0;
            m_disposed = false;
        }

        public void setClientDispose()
        {

        }

        public bool getClientDispose()
        {
            return false;
        }

        public void setLuaFunctor(LuaTable luaTable, LuaFunction function)
        {
            m_timerDisp.setLuaFunctor(luaTable, function);
        }
    }
}